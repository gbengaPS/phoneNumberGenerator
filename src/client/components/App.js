import React, { Component } from 'react';
import axios from 'axios';
import '../style.css';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      total: '',
      min: '',
      max: '',
      numbers: [],
      error: '',
      numberInput: null,
      numbersFetched: false,
    };
  }

  componentDidMount() {
    axios
      .get('/api/v1/numbers')
      .then(({ data }) => {
        this.updateAppState(data);
      })
      .catch((error) => {
        this.setState({ error });
      });
  }

  updateAppState = (data) => {
    const {
      total, min, max, numbers,
    } = data;
    this.setState({
      total,
      min,
      max,
      numbers,
      numbersFetched: true,
    });
  };

  handleChange = (event) => {
    event.preventDefault();
    this.setState({ numberInput: event.target.value });
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    this.setState({ numbersFetched: false });
    const body = { count: this.state.numberInput };
    try {
      await axios.post('/api/v1/numbers/generate', body);
      const response = await axios.get('/api/v1/numbers');
      this.updateAppState(response.data);
    } catch (error) {
      this.setState({ error: 'There was an error completing the request' });
    }
  };

  renderForm = () => (
    <form onSubmit={this.handleSubmit} className="form">
      <input
        type="number"
        className="input"
        onChange={this.handleChange}
        required
      />
      <input type="submit" className="btn" value="Generate Numbers" />
    </form>
  );

  renderTable = numbers => (
    <table className="card">
      <thead>
        <tr className="text-bold">
          <th>SN</th>
          <th>Number</th>
        </tr>
      </thead>
      <tbody>
        {numbers.map((number, index) => (
          <tr key={index + 1}>
            <td>{index + 1}</td>
            <td>{number}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );

  renderLoader = () => <p className="center">Fetching numbers...</p>;

  render() {
    const {
      max, min, total, numbers, numbersFetched,
    } = this.state;
    return (
      <div className="wrapper">
        <div className="side-bar">{this.renderForm()}</div>
        <div className="main-section">
          <h2>Phone Numbers</h2>
          {numbersFetched ? this.renderTable(numbers) : this.renderLoader()}
        </div>
        <div className="stats">
          <h2>Stats</h2>
          <ul className="card">
            <li>
              <span className="text-bold">Total Numbers:</span> {total}
            </li>
            <li>
              <span className="text-bold">Min Number:</span> {min}
            </li>
            <li>
              <span className="text-bold">Max Number:</span> {max}
            </li>
          </ul>
        </div>
        <div className="clear" />
      </div>
    );
  }
}
export default App;
