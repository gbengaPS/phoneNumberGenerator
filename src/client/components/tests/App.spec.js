import React from 'react';
import { mount } from 'enzyme';
import * as axios from 'axios';
import App from '../App';

jest.mock('axios');
axios.get.mockImplementation(() => Promise.resolve({ data: { numbers: ['234'] } }));
const wrapper = mount(<App />);
describe('App', () => {
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state when handleChange is called', () => {
    const event = { target: { value: '34' }, preventDefault: jest.fn() };
    wrapper.instance().handleChange(event);
    const state = wrapper.state();
    expect(state.numberInput).toBe(event.target.value);
  });

  it('should set state when handleSubmit is called', () => {
    const event = { preventDefault: jest.fn() };
    wrapper.instance().handleSubmit(event);
    wrapper.setState({ numberInput: '90' });
    const state = wrapper.state();
    expect(state.numbersFetched).toBe(false);
  });
});
