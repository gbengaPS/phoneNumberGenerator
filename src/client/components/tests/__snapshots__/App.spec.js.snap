// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<App>
  <div
    className="wrapper"
  >
    <div
      className="side-bar"
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <input
          className="input"
          onChange={[Function]}
          required={true}
          type="number"
        />
        <input
          className="btn"
          type="submit"
          value="Generate Numbers"
        />
      </form>
    </div>
    <div
      className="main-section"
    >
      <h2>
        Phone Numbers
      </h2>
      <p
        className="center"
      >
        Fetching numbers...
      </p>
    </div>
    <div
      className="stats"
    >
      <h2>
        Stats
      </h2>
      <ul
        className="card"
      >
        <li>
          <span
            className="text-bold"
          >
            Total Numbers:
          </span>
           
        </li>
        <li>
          <span
            className="text-bold"
          >
            Min Number:
          </span>
           
        </li>
        <li>
          <span
            className="text-bold"
          >
            Max Number:
          </span>
           
        </li>
      </ul>
    </div>
    <div
      className="clear"
    />
  </div>
</App>
`;
